python/lambda-ddb-mysql-etl-pipeline/lambda/dbwrite.py

```diff
@@ -0,0 +1,200 @@
+_lambda = boto3.client('lambda')
+db_conn = None
+SCHEMA = os.environ['SCHEMA']
+
+def handler(event, context):
+    """Retrieve s3 object and write to dynamo or mysql table(not-functional in this example, thus comments)
+
+    Args:
+        event (dict, list, str, int, float, NoneType): Provides event specific data to the handler. It
+            can also be of type list, str, int, float, or NoneType.
+        context (obj): Provides methods and properties that contain invocation, function and
+            execution environment information.
+    Raises:
+        Exception: If file format is not of .json or .csv file types.
+
+    Notes:
+        A call to `write_to_mysql()` will require that you setup a DB Connection
+        and set `db_conn` herein to `True`. Otherwise, the current example will
+        only print results to CloudWatch Logs. Pandas & PyMySQL archives are
+        included as layers in the project.
+    """
+    s3_key=event['Records'][0]['s3']['object']['key']
+    stream_obj=s3.get_object(Bucket=os.environ['S3_BUCKET'], Key=s3_key)
+    data = stream_obj['Body'].read().decode('utf-8')
+    if mimetypes.guess_type(s3_key)[0] == "application/json":
+        asteroid_list=json.loads(data, parse_float=decimal.Decimal)
+        write_to_ddb(asteroid_list)
+    elif mimetypes.guess_type(s3_key)[0] ==  "text/csv":
+        write_to_mysql(data, 'asteroids')
+    else:
+        raise Exception("FILE FORMAT NOT RECOGNIZED")
+
+def write_to_ddb(ast_list):
+    """Puts/Writes each JSON record to DynamoDB.
+
+    Args:
+        ast_list (obj): A JSON list of objects that contain Asteroid data.
+    """
+    for index in ast_list:
+        ddb_table.put_item(Item=index)
+
+def write_to_mysql(ast_list, t_suffix):
+    mysql_attrs=get_mysql_attrs(t_suffix)
+    dynamic_mysql_crud_ops(ast_list, t_suffix, mysql_attrs)
+
+def get_mysql_attrs(t_suffix):
+    """Gets table attributes from Information Schema & creates new staging table, if necessary
+
+    Args:
+         t_suffix (str): The table name suffix of the write target table.
+    Returns:
+        attribute_list (list): Array of column names in the specified db table.
+    """
+    if db_conn is None:
+        return "`mysql_attrs` ref from `get_mysql_attrs()`"
+    else:
+        # Gets DB Creds from AWS Secrets Manager
+        try:
+            session = boto3.session.Session()
+            client = session.client(service_name='secretsmanager'
+                , region_name=os.environ['REGION'])
+            SECRET = client.get_secret_value(SecretId=os.environ['DB_SECRETS_REF'])
+            if 'SecretString' in SECRET:
+                SECRETS = json.loads(SECRET['SecretString'])
+            else:
+                SECRETS = json.loads(b64decode(SECRET['SecretBinary']))
+        except Exception:
+            logger.error("ERROR: Unable to GET DB Credentials from Secrets Manager")
+
+        try:
+            connection = pymysql.connect(host=SECRETS['MYSQL_ENDPOINT'], port=3306
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *connection*. The resource is allocated by call *pymysql.connect*. Execution paths that do not contain closure statements were detected. Close *connection* in a try-finally block to prevent the resource leak.